grpc-cronet: ExperimentalBidirectionalStream, ExperimentalCronetEngine are deprecated

I Attempted updating cronet dependencies in #10359:

```
org.chromium.net:cronet-api 108.5359.79 -> 113.5672.61
org.chromium.net:cronet-embedded 108.5359.79 -> 113.5672.61
```

Gradle build failed with various deprecation warnings:

<details>
<summary>gradle build log</summary>

(shortened repeated warnings)

```java
> Task :grpc-cronet:compileDebugJavaWithJavac
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:50: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
import org.chromium.net.ExperimentalBidirectionalStream;
                       ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:51: warning: [deprecation] ExperimentalCronetEngine in org.chromium.net has been deprecated
import org.chromium.net.ExperimentalCronetEngine;
                       ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:58: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
import org.chromium.net.ExperimentalBidirectionalStream;
                       ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:320: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
      ExperimentalBidirectionalStream.Builder builder =
      ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:320: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
      ExperimentalBidirectionalStream.Builder builder =
                                     ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:321: warning: [deprecation] ExperimentalCronetEngine in org.chromium.net has been deprecated
          ((ExperimentalCronetEngine) cronetEngine)
            ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:332: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
    private static void setTrafficStatsTag(ExperimentalBidirectionalStream.Builder builder,
                                           ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:332: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
    private static void setTrafficStatsTag(ExperimentalBidirectionalStream.Builder builder,
                                                                          ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:338: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
              setTrafficStatsTagMethod = ExperimentalBidirectionalStream.Builder.class
                                         ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:338: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
              setTrafficStatsTagMethod = ExperimentalBidirectionalStream.Builder.class
                                                                        ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:361: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
    private static void setTrafficStatsUid(ExperimentalBidirectionalStream.Builder builder,
                                           ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:361: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
    private static void setTrafficStatsUid(ExperimentalBidirectionalStream.Builder builder,
                                                                          ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:367: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
              setTrafficStatsUidMethod = ExperimentalBidirectionalStream.Builder.class
                                         ^
cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java:367: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
              setTrafficStatsUidMethod = ExperimentalBidirectionalStream.Builder.class
                                                                        ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:196: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
        ExperimentalBidirectionalStream.Builder expBidiStreamBuilder =
        ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:196: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
        ExperimentalBidirectionalStream.Builder expBidiStreamBuilder =
                                       ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:197: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
            (ExperimentalBidirectionalStream.Builder) builder;
             ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:197: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
            (ExperimentalBidirectionalStream.Builder) builder;
                                            ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:369: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
  private static void addRequestAnnotation(ExperimentalBidirectionalStream.Builder builder,
                                           ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:369: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
  private static void addRequestAnnotation(ExperimentalBidirectionalStream.Builder builder,
                                                                          ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:375: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
            addRequestAnnotationMethod = ExperimentalBidirectionalStream.Builder.class
                                         ^
cronet/src/main/java/io/grpc/cronet/CronetClientStream.java:375: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
            addRequestAnnotationMethod = ExperimentalBidirectionalStream.Builder.class
                                                                        ^
22 warnings

> Task :grpc-cronet:compileDebugUnitTestJavaWithJavac
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:57: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
import org.chromium.net.ExperimentalBidirectionalStream;
                       ^
cronet/src/test/java/io/grpc/cronet/CronetChannelBuilderTest.java:38: warning: [deprecation] ExperimentalCronetEngine in org.chromium.net has been deprecated
import org.chromium.net.ExperimentalCronetEngine;
                       ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:79: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
  @Mock private ExperimentalBidirectionalStream cronetStream;
                ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:81: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
  @Mock private ExperimentalBidirectionalStream.Builder builder;
                ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:81: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
  @Mock private ExperimentalBidirectionalStream.Builder builder;
                                               ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:684: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
    ExperimentalBidirectionalStream.Builder getBuilder =
    ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:684: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
    ExperimentalBidirectionalStream.Builder getBuilder =
                                   ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:685: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
        mock(ExperimentalBidirectionalStream.Builder.class);
             ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:685: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
        mock(ExperimentalBidirectionalStream.Builder.class);
                                            ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:741: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
    ExperimentalBidirectionalStream.Builder builder =
    ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:741: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
    ExperimentalBidirectionalStream.Builder builder =
                                   ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:742: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
        mock(ExperimentalBidirectionalStream.Builder.class);
             ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:742: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
        mock(ExperimentalBidirectionalStream.Builder.class);
                                            ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:773: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
    ExperimentalBidirectionalStream.Builder builder =
    ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:773: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
    ExperimentalBidirectionalStream.Builder builder =
                                   ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:774: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
        mock(ExperimentalBidirectionalStream.Builder.class);
             ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:774: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
        mock(ExperimentalBidirectionalStream.Builder.class);
                                            ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:813: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
    ExperimentalBidirectionalStream.Builder builder =
    ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:813: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
    ExperimentalBidirectionalStream.Builder builder =
                                   ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:814: warning: [deprecation] ExperimentalBidirectionalStream in org.chromium.net has been deprecated
        mock(ExperimentalBidirectionalStream.Builder.class);
             ^
cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java:814: warning: [deprecation] Builder in ExperimentalBidirectionalStream has been deprecated
        mock(ExperimentalBidirectionalStream.Builder.class);
                                            ^
cronet/src/test/java/io/grpc/cronet/CronetChannelBuilderTest.java:53: warning: [deprecation] ExperimentalCronetEngine in org.chromium.net has been deprecated
  @Mock private ExperimentalCronetEngine mockEngine;
                ^
22 warnings

```
</details>

### Steps to reproduce the bug

Apply patch:
```
diff --git a/gradle/libs.versions.toml b/gradle/libs.versions.toml
index 7217d560a..b022132a9 100644
--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -36,8 +36,8 @@ auto-value-annotations = "com.google.auto.value:auto-value-annotations:1.10.2"
 checkstyle = "com.puppycrawl.tools:checkstyle:10.12.1"
 commons-math3 = "org.apache.commons:commons-math3:3.6.1"
 conscrypt = "org.conscrypt:conscrypt-openjdk-uber:2.5.2"
-cronet-api = "org.chromium.net:cronet-api:108.5359.79"
-cronet-embedded = "org.chromium.net:cronet-embedded:108.5359.79"
+cronet-api = "org.chromium.net:cronet-api:113.5672.61"
+cronet-embedded = "org.chromium.net:cronet-embedded:113.5672.61"
 errorprone-annotations = "com.google.errorprone:error_prone_annotations:2.20.0"
 errorprone-core = "com.google.errorprone:error_prone_core:2.20.0"
 google-api-protos = "com.google.api.grpc:proto-google-common-protos:2.22.0"
```

Then do `gradle build`.

--- a/cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java
+++ b/cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java
@@ -84,7 +84,7 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
 
   private final CronetEngine cronetEngine;
   private final ManagedChannelImplBuilder managedChannelImplBuilder;
-  private TransportTracer.Factory transportTracerFactory = TransportTracer.getDefaultFactory();
+  private final TransportTracer.Factory transportTracerFactory = TransportTracer.getDefaultFactory();
 
   private boolean alwaysUsePut = false;
 
@@ -132,7 +132,7 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
    * Sets the maximum message size allowed to be received on the channel. If not called,
    * defaults to {@link io.grpc.internal.GrpcUtil#DEFAULT_MAX_MESSAGE_SIZE}.
    */
-  public final CronetChannelBuilder maxMessageSize(int maxMessageSize) {
+  public CronetChannelBuilder maxMessageSize(int maxMessageSize) {
     checkArgument(maxMessageSize >= 0, "maxMessageSize must be >= 0");
     this.maxMessageSize = maxMessageSize;
     return this;
@@ -141,7 +141,7 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
   /**
    * Sets the Cronet channel to always use PUT instead of POST. Defaults to false.
    */
-  public final CronetChannelBuilder alwaysUsePut(boolean enable) {
+  public CronetChannelBuilder alwaysUsePut(boolean enable) {
     this.alwaysUsePut = enable;
     return this;
   }
@@ -163,7 +163,7 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
    *     application.
    * @return the builder to facilitate chaining.
    */
-  final CronetChannelBuilder setTrafficStatsTag(int tag) {
+  CronetChannelBuilder setTrafficStatsTag(int tag) {
     trafficStatsTagSet = true;
     trafficStatsTag = tag;
     return this;
@@ -184,7 +184,7 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
    * @param uid the UID to attribute socket traffic caused by this channel.
    * @return the builder to facilitate chaining.
    */
-  final CronetChannelBuilder setTrafficStatsUid(int uid) {
+  CronetChannelBuilder setTrafficStatsUid(int uid) {
     trafficStatsUidSet = true;
     trafficStatsUid = uid;
     return this;
@@ -200,7 +200,7 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
    *
    * @since 1.12.0
    */
-  public final CronetChannelBuilder scheduledExecutorService(
+  public CronetChannelBuilder scheduledExecutorService(
       ScheduledExecutorService scheduledExecutorService) {
     this.scheduledExecutorService =
         checkNotNull(scheduledExecutorService, "scheduledExecutorService");

--- a/cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java
+++ b/cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java
@@ -20,7 +20,6 @@ import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
 import static io.grpc.internal.GrpcUtil.DEFAULT_MAX_MESSAGE_SIZE;
 
-import android.util.Log;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
 import com.google.common.util.concurrent.MoreExecutors;
@@ -38,8 +37,6 @@ import io.grpc.internal.ManagedChannelImplBuilder;
 import io.grpc.internal.ManagedChannelImplBuilder.ClientTransportFactoryBuilder;
 import io.grpc.internal.SharedResourceHolder;
 import io.grpc.internal.TransportTracer;
-import java.lang.reflect.InvocationTargetException;
-import java.lang.reflect.Method;
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 import java.util.Collection;
@@ -49,14 +46,10 @@ import java.util.concurrent.ScheduledExecutorService;
 import javax.annotation.Nullable;
 import org.chromium.net.BidirectionalStream;
 import org.chromium.net.CronetEngine;
-import org.chromium.net.ExperimentalBidirectionalStream;
-import org.chromium.net.ExperimentalCronetEngine;
 
 /** Convenience class for building channels with the cronet transport. */
 @ExperimentalApi("There is no plan to make this API stable, given transport API instability")
 public final class CronetChannelBuilder extends ForwardingChannelBuilder2<CronetChannelBuilder> {
-
-  private static final String LOG_TAG = "CronetChannelBuilder";
 
   /** BidirectionalStream.Builder factory used for getting the gRPC BidirectionalStream. */
   public static abstract class StreamBuilderFactory {
@@ -296,11 +289,6 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
    * StreamBuilderFactory impl that applies TrafficStats tags to stream builders that are produced.
    */
   private static class TaggingStreamFactory extends StreamBuilderFactory {
-    private static volatile boolean loadSetTrafficStatsTagAttempted;
-    private static volatile boolean loadSetTrafficStatsUidAttempted;
-    private static volatile Method setTrafficStatsTagMethod;
-    private static volatile Method setTrafficStatsUidMethod;
-
     private final CronetEngine cronetEngine;
     private final boolean trafficStatsTagSet;
     private final int trafficStatsTag;
@@ -323,74 +311,16 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
     @Override
     public BidirectionalStream.Builder newBidirectionalStreamBuilder(
         String url, BidirectionalStream.Callback callback, Executor executor) {
-      ExperimentalBidirectionalStream.Builder builder =
-          ((ExperimentalCronetEngine) cronetEngine)
+      BidirectionalStream.Builder builder =
+          cronetEngine
               .newBidirectionalStreamBuilder(url, callback, executor);
       if (trafficStatsTagSet) {
-        setTrafficStatsTag(builder, trafficStatsTag);
+        builder.setTrafficStatsTag(trafficStatsTag);
       }
       if (trafficStatsUidSet) {
-        setTrafficStatsUid(builder, trafficStatsUid);
+        builder.setTrafficStatsUid(trafficStatsUid);
       }
       return builder;
-    }
-
-    private static void setTrafficStatsTag(ExperimentalBidirectionalStream.Builder builder,
-        int tag) {
-      if (!loadSetTrafficStatsTagAttempted) {
-        synchronized (TaggingStreamFactory.class) {
-          if (!loadSetTrafficStatsTagAttempted) {
-            try {
-              setTrafficStatsTagMethod = ExperimentalBidirectionalStream.Builder.class
-                  .getMethod("setTrafficStatsTag", int.class);
-            } catch (NoSuchMethodException e) {
-              Log.w(LOG_TAG,
-                  "Failed to load method ExperimentalBidirectionalStream.Builder.setTrafficStatsTag",
-                  e);
-            } finally {
-              loadSetTrafficStatsTagAttempted = true;
-            }
-          }
-        }
-      }
-      if (setTrafficStatsTagMethod != null) {
-        try {
-          setTrafficStatsTagMethod.invoke(builder, tag);
-        } catch (InvocationTargetException e) {
-          throw new RuntimeException(e.getCause() == null ? e.getTargetException() : e.getCause());
-        } catch (IllegalAccessException e) {
-          Log.w(LOG_TAG, "Failed to set traffic stats tag: " + tag, e);
-        }
-      }
-    }
-
-    private static void setTrafficStatsUid(ExperimentalBidirectionalStream.Builder builder,
-        int uid) {
-      if (!loadSetTrafficStatsUidAttempted) {
-        synchronized (TaggingStreamFactory.class) {
-          if (!loadSetTrafficStatsUidAttempted) {
-            try {
-              setTrafficStatsUidMethod = ExperimentalBidirectionalStream.Builder.class
-                  .getMethod("setTrafficStatsUid", int.class);
-            } catch (NoSuchMethodException e) {
-              Log.w(LOG_TAG,
-                  "Failed to load method ExperimentalBidirectionalStream.Builder.setTrafficStatsUid",
-                  e);
-            } finally {
-              loadSetTrafficStatsUidAttempted = true;
-            }
-          }
-        }
-      }
-      if (setTrafficStatsUidMethod != null) {
-        try {
-          setTrafficStatsUidMethod.invoke(builder, uid);
-        } catch (InvocationTargetException e) {
-          throw new RuntimeException(e.getCause() == null ? e.getTargetException() : e.getCause());
-        } catch (IllegalAccessException e) {
-          Log.w(LOG_TAG, "Failed to set traffic stats uid: " + uid, e);
-        }
-      }
     }
   }
 }

--- a/cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java
+++ b/cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java
@@ -173,7 +173,7 @@ public final class CronetChannelBuilder extends ForwardingChannelBuilder2<Cronet
    * Sets specific UID to use when accounting socket traffic caused by this channel. See {@link
    * android.net.TrafficStats} for more information. Designed for use when performing an operation
    * on behalf of another application. Caller must hold {@link
-   * android.Manifest.permission#MODIFY_NETWORK_ACCOUNTING} permission. By default traffic is
+   * android.Manifest.permission#UPDATE_DEVICE_STATS} permission. By default traffic is
    * attributed to UID of caller.
    *
    * <p><b>NOTE:</b>Setting a UID disallows sharing of sockets with channels with other UIDs, which

--- a/cronet/src/main/java/io/grpc/cronet/CronetClientStream.java
+++ b/cronet/src/main/java/io/grpc/cronet/CronetClientStream.java
@@ -247,7 +247,7 @@ class CronetClientStream extends AbstractClientStream {
   class TransportState extends Http2ClientStreamTransportState {
     private final Object lock;
     @GuardedBy("lock")
-    private Collection<PendingData> pendingData = new ArrayList<PendingData>();
+    private final Collection<PendingData> pendingData = new ArrayList<>();
     @GuardedBy("lock")
     private boolean streamReady;
     @GuardedBy("lock")

--- a/cronet/src/main/java/io/grpc/cronet/CronetClientStream.java
+++ b/cronet/src/main/java/io/grpc/cronet/CronetClientStream.java
@@ -40,8 +40,6 @@ import io.grpc.internal.StatsTraceContext;
 import io.grpc.internal.TransportFrameUtil;
 import io.grpc.internal.TransportTracer;
 import io.grpc.internal.WritableBuffer;
-import java.lang.reflect.InvocationTargetException;
-import java.lang.reflect.Method;
 import java.nio.Buffer;
 import java.nio.ByteBuffer;
 import java.nio.charset.Charset;
@@ -55,7 +53,6 @@ import javax.annotation.Nullable;
 import javax.annotation.concurrent.GuardedBy;
 import org.chromium.net.BidirectionalStream;
 import org.chromium.net.CronetException;
-import org.chromium.net.ExperimentalBidirectionalStream;
 import org.chromium.net.UrlResponseInfo;
 
 /**
@@ -65,9 +62,6 @@ class CronetClientStream extends AbstractClientStream {
   private static final int READ_BUFFER_CAPACITY = 4 * 1024;
   private static final ByteBuffer EMPTY_BUFFER = ByteBuffer.allocateDirect(0);
   private static final String LOG_TAG = "grpc-java-cronet";
-
-  private static volatile boolean loadAddRequestAnnotationAttempted;
-  private static volatile Method addRequestAnnotationMethod;
 
   @Deprecated
   static final CallOptions.Key<Object> CRONET_ANNOTATION_KEY =
@@ -194,14 +188,12 @@ class CronetClientStream extends AbstractClientStream {
         builder.delayRequestHeadersUntilFirstFlush(true);
       }
       if (annotation != null || annotations != null) {
-        ExperimentalBidirectionalStream.Builder expBidiStreamBuilder =
-            (ExperimentalBidirectionalStream.Builder) builder;
         if (annotation != null) {
-          addRequestAnnotation(expBidiStreamBuilder, annotation);
+          builder.addRequestAnnotation(annotation);
         }
         if (annotations != null) {
           for (Object o : annotations) {
-            addRequestAnnotation(expBidiStreamBuilder, o);
+            builder.addRequestAnnotation(o);
           }
         }
       }
@@ -365,35 +357,6 @@ class CronetClientStream extends AbstractClientStream {
     return !CONTENT_TYPE_KEY.name().equalsIgnoreCase(key)
         && !USER_AGENT_KEY.name().equalsIgnoreCase(key)
         && !TE_HEADER.name().equalsIgnoreCase(key);
-  }
-
-  private static void addRequestAnnotation(ExperimentalBidirectionalStream.Builder builder,
-      Object annotation) {
-    if (!loadAddRequestAnnotationAttempted) {
-      synchronized (CronetClientStream.class) {
-        if (!loadAddRequestAnnotationAttempted) {
-          try {
-            addRequestAnnotationMethod = ExperimentalBidirectionalStream.Builder.class
-                .getMethod("addRequestAnnotation", Object.class);
-          } catch (NoSuchMethodException e) {
-            Log.w(LOG_TAG,
-                "Failed to load method ExperimentalBidirectionalStream.Builder.addRequestAnnotation",
-                e);
-          } finally {
-            loadAddRequestAnnotationAttempted = true;
-          }
-        }
-      }
-    }
-    if (addRequestAnnotationMethod != null) {
-      try {
-        addRequestAnnotationMethod.invoke(builder, annotation);
-      } catch (InvocationTargetException e) {
-        throw new RuntimeException(e.getCause() == null ? e.getTargetException() : e.getCause());
-      } catch (IllegalAccessException e) {
-        Log.w(LOG_TAG, "Failed to add request annotation: " + annotation, e);
-      }
-    }
   }
 
   private void setGrpcHeaders(BidirectionalStream.Builder builder) {

--- a/cronet/src/main/java/io/grpc/cronet/CronetClientStream.java
+++ b/cronet/src/main/java/io/grpc/cronet/CronetClientStream.java
@@ -42,7 +42,7 @@ import io.grpc.internal.TransportTracer;
 import io.grpc.internal.WritableBuffer;
 import java.nio.Buffer;
 import java.nio.ByteBuffer;
-import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -372,10 +372,10 @@ class CronetClientStream extends AbstractClientStream {
     // String and byte array.
     byte[][] serializedHeaders = TransportFrameUtil.toHttp2Headers(headers);
     for (int i = 0; i < serializedHeaders.length; i += 2) {
-      String key = new String(serializedHeaders[i], Charset.forName("UTF-8"));
+      String key = new String(serializedHeaders[i], StandardCharsets.UTF_8);
       // TODO(ericgribkoff): log an error or throw an exception
       if (isApplicationHeader(key)) {
-        String value = new String(serializedHeaders[i + 1], Charset.forName("UTF-8"));
+        String value = new String(serializedHeaders[i + 1], StandardCharsets.UTF_8);
         builder.addHeader(key, value);
       }
     }
@@ -552,8 +552,8 @@ class CronetClientStream extends AbstractClientStream {
 
       byte[][] headerValues = new byte[headerList.size()][];
       for (int i = 0; i < headerList.size(); i += 2) {
-        headerValues[i] = headerList.get(i).getBytes(Charset.forName("UTF-8"));
-        headerValues[i + 1] = headerList.get(i + 1).getBytes(Charset.forName("UTF-8"));
+        headerValues[i] = headerList.get(i).getBytes(StandardCharsets.UTF_8);
+        headerValues[i + 1] = headerList.get(i + 1).getBytes(StandardCharsets.UTF_8);
       }
       Metadata metadata =
           InternalMetadata.newMetadata(TransportFrameUtil.toRawSerializedHeaders(headerValues));

--- a/cronet/src/main/java/io/grpc/cronet/CronetClientTransport.java
+++ b/cronet/src/main/java/io/grpc/cronet/CronetClientTransport.java
@@ -56,7 +56,7 @@ class CronetClientTransport implements ConnectionClientTransport {
   private final Object lock = new Object();
   @GuardedBy("lock")
   private final Set<CronetClientStream> streams = Collections.newSetFromMap(
-          new IdentityHashMap<CronetClientStream, Boolean>());
+          new IdentityHashMap<>());
   private final Executor executor;
   private final int maxMessageSize;
   private final boolean alwaysUsePut;
@@ -64,6 +64,7 @@ class CronetClientTransport implements ConnectionClientTransport {
   private Attributes attrs;
   private final boolean useGetForSafeMethods;
   private final boolean usePutForIdempotentMethods;
+  private final StreamBuilderFactory streamFactory;
   // Indicates the transport is in go-away state: no new streams will be processed,
   // but existing streams may continue.
   @GuardedBy("lock")
@@ -79,7 +80,6 @@ class CronetClientTransport implements ConnectionClientTransport {
   @GuardedBy("lock")
   // Whether this transport has started.
   private boolean started;
-  private StreamBuilderFactory streamFactory;
 
   CronetClientTransport(
       StreamBuilderFactory streamFactory,
@@ -205,9 +205,9 @@ class CronetClientTransport implements ConnectionClientTransport {
       // streams.remove()
       streamsCopy = new ArrayList<>(streams);
     }
-    for (int i = 0; i < streamsCopy.size(); i++) {
+    for (CronetClientStream cronetClientStream : streamsCopy) {
       // Avoid deadlock by calling into stream without lock held
-      streamsCopy.get(i).cancel(status);
+      cronetClientStream.cancel(status);
     }
     stopIfNecessary();
   }
@@ -255,7 +255,7 @@ class CronetClientTransport implements ConnectionClientTransport {
    */
   void stopIfNecessary() {
     synchronized (lock) {
-      if (goAway && !stopped && streams.size() == 0) {
+      if (goAway && !stopped && streams.isEmpty()) {
         stopped = true;
       } else {
         return;

--- a/cronet/src/main/java/io/grpc/cronet/InternalCronetChannelBuilder.java
+++ b/cronet/src/main/java/io/grpc/cronet/InternalCronetChannelBuilder.java
@@ -47,7 +47,7 @@ public final class InternalCronetChannelBuilder {
    * Sets specific UID to use when accounting socket traffic caused by this channel. See {@link
    * android.net.TrafficStats} for more information. Designed for use when performing an operation
    * on behalf of another application. Caller must hold {@link
-   * android.Manifest.permission#MODIFY_NETWORK_ACCOUNTING} permission. By default traffic is
+   * android.Manifest.permission#UPDATE_DEVICE_STATS} permission. By default traffic is
    * attributed to UID of caller.
    *
    * <p><b>NOTE:</b>Setting a UID disallows sharing of sockets with channels with other UIDs, which

--- a/cronet/src/test/java/io/grpc/cronet/CronetChannelBuilderTest.java
+++ b/cronet/src/test/java/io/grpc/cronet/CronetChannelBuilderTest.java
@@ -35,7 +35,7 @@ import io.grpc.internal.SharedResourceHolder;
 import io.grpc.testing.TestMethodDescriptors;
 import java.net.InetSocketAddress;
 import java.util.concurrent.ScheduledExecutorService;
-import org.chromium.net.ExperimentalCronetEngine;
+import org.chromium.net.CronetEngine;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -50,7 +50,7 @@ import org.robolectric.annotation.Config;
 public final class CronetChannelBuilderTest {
   @Rule public final MockitoRule mocks = MockitoJUnit.rule();
 
-  @Mock private ExperimentalCronetEngine mockEngine;
+  @Mock private CronetEngine mockEngine;
   @Mock private ChannelLogger channelLogger;
 
   private final ClientStreamTracer[] tracers =

--- a/cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java
+++ b/cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java
@@ -80,12 +80,7 @@ public final class CronetClientStreamTest {
   @Mock private BidirectionalStream.Builder builder;
   private final Object lock = new Object();
   private final TransportTracer transportTracer = TransportTracer.getDefaultFactory().create();
-  private final Executor executor = new Executor() {
-      @Override
-      public void execute(Runnable r) {
-        r.run();
-      }
-    };
+  private final Executor executor = Runnable::run;
   CronetClientStream clientStream;
 
   private MethodDescriptor.Marshaller<Void> marshaller = TestMethodDescriptors.voidMarshaller();
@@ -171,7 +166,7 @@ public final class CronetClientStreamTest {
     String[] requests = new String[5];
     WritableBuffer[] buffers = new WritableBuffer[5];
     for (int i = 0; i < 5; ++i) {
-      requests[i] = new String("request" + String.valueOf(i));
+      requests[i] = "request" + i;
       buffers[i] = allocator.allocate(requests[i].length());
       buffers[i].write(requests[i].getBytes(Charset.forName("UTF-8")), 0, requests[i].length());
       // The 3rd and 5th writeFrame calls have flush=true.
@@ -206,27 +201,19 @@ public final class CronetClientStreamTest {
   }
 
   private static List<Map.Entry<String, String>> responseHeader(String status) {
-    Map<String, String> headers = new HashMap<String, String>();
+    Map<String, String> headers = new HashMap<>();
     headers.put(":status", status);
     headers.put("content-type", "application/grpc");
     headers.put("test-key", "test-value");
-    List<Map.Entry<String, String>> headerList = new ArrayList<Map.Entry<String, String>>(3);
-    for (Map.Entry<String, String> entry : headers.entrySet()) {
-      headerList.add(entry);
-    }
-    return headerList;
+    return new ArrayList<>(headers.entrySet());
   }
 
   private static List<Map.Entry<String, String>> trailers(int status) {
-    Map<String, String> trailers = new HashMap<String, String>();
+    Map<String, String> trailers = new HashMap<>();
     trailers.put("grpc-status", String.valueOf(status));
     trailers.put("content-type", "application/grpc");
     trailers.put("test-trailer-key", "test-trailer-value");
-    List<Map.Entry<String, String>> trailerList = new ArrayList<Map.Entry<String, String>>(3);
-    for (Map.Entry<String, String> entry : trailers.entrySet()) {
-      trailerList.add(entry);
-    }
-    return trailerList;
+    return new ArrayList<>(trailers.entrySet());
   }
 
   private static ByteBuffer createMessageFrame(byte[] bytes) {

--- a/cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java
+++ b/cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java
@@ -18,6 +18,7 @@ package io.grpc.cronet;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
@@ -46,7 +47,7 @@ import io.grpc.testing.TestMethodDescriptors;
 import java.io.ByteArrayInputStream;
 import java.nio.Buffer;
 import java.nio.ByteBuffer;
-import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
@@ -102,7 +103,7 @@ public final class CronetClientStreamTest {
     @Override
     @SuppressWarnings("GuardedBy")
     public void run() {
-      assertTrue(stream != null);
+      assertNotNull(stream);
       stream.transportState().start(factory);
     }
   }
@@ -168,7 +169,7 @@ public final class CronetClientStreamTest {
     for (int i = 0; i < 5; ++i) {
       requests[i] = "request" + i;
       buffers[i] = allocator.allocate(requests[i].length());
-      buffers[i].write(requests[i].getBytes(Charset.forName("UTF-8")), 0, requests[i].length());
+      buffers[i].write(requests[i].getBytes(StandardCharsets.UTF_8), 0, requests[i].length());
       // The 3rd and 5th writeFrame calls have flush=true.
       clientStream.abstractClientStreamSink().writeFrame(buffers[i], false, i == 2 || i == 4, 1);
     }
@@ -253,7 +254,7 @@ public final class CronetClientStreamTest {
     callback.onReadCompleted(
         cronetStream,
         info,
-        createMessageFrame(new String("response1").getBytes(Charset.forName("UTF-8"))),
+        createMessageFrame("response1".getBytes(StandardCharsets.UTF_8)),
         false);
     // Haven't request any message, so no callback is called here.
     verify(clientListener, times(0)).messagesAvailable(isA(MessageProducer.class));
@@ -283,9 +284,9 @@ public final class CronetClientStreamTest {
     verify(cronetStream, times(0)).write(isA(ByteBuffer.class), isA(Boolean.class));
     // Send the first data frame.
     CronetWritableBufferAllocator allocator = new CronetWritableBufferAllocator();
-    String request = new String("request");
+    String request = "request";
     WritableBuffer writableBuffer = allocator.allocate(request.length());
-    writableBuffer.write(request.getBytes(Charset.forName("UTF-8")), 0, request.length());
+    writableBuffer.write(request.getBytes(StandardCharsets.UTF_8), 0, request.length());
     clientStream.abstractClientStreamSink().writeFrame(writableBuffer, false, true, 1);
     ArgumentCaptor<ByteBuffer> bufferCaptor = ArgumentCaptor.forClass(ByteBuffer.class);
     verify(cronetStream, times(1)).write(bufferCaptor.capture(), isA(Boolean.class));
@@ -304,7 +305,7 @@ public final class CronetClientStreamTest {
     callback.onReadCompleted(
         cronetStream,
         info,
-        createMessageFrame(new String("response").getBytes(Charset.forName("UTF-8"))),
+        createMessageFrame("response".getBytes(StandardCharsets.UTF_8)),
         false);
     verify(clientListener, times(1)).messagesAvailable(isA(MessageProducer.class));
     verify(cronetStream, times(2)).read(isA(ByteBuffer.class));
@@ -680,7 +681,7 @@ public final class CronetClientStreamTest {
         .newBidirectionalStreamBuilder(
             isA(String.class), isA(BidirectionalStream.Callback.class), isA(Executor.class));
 
-    byte[] msg = "request".getBytes(Charset.forName("UTF-8"));
+    byte[] msg = "request".getBytes(StandardCharsets.UTF_8);
     stream.writeMessage(new ByteArrayInputStream(msg));
     // We still haven't built the stream or sent anything.
     verify(cronetStream, times(0)).write(isA(ByteBuffer.class), isA(Boolean.class));

--- a/cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java
+++ b/cronet/src/test/java/io/grpc/cronet/CronetClientStreamTest.java
@@ -54,7 +54,6 @@ import java.util.Map;
 import java.util.concurrent.Executor;
 import org.chromium.net.BidirectionalStream;
 import org.chromium.net.CronetException;
-import org.chromium.net.ExperimentalBidirectionalStream;
 import org.chromium.net.UrlResponseInfo;
 import org.chromium.net.impl.UrlResponseInfoImpl;
 import org.junit.Before;
@@ -76,9 +75,9 @@ public final class CronetClientStreamTest {
   @Mock private CronetClientTransport transport;
   private Metadata metadata = new Metadata();
   @Mock private StreamBuilderFactory factory;
-  @Mock private ExperimentalBidirectionalStream cronetStream;
+  @Mock private BidirectionalStream cronetStream;
   @Mock private ClientStreamListener clientListener;
-  @Mock private ExperimentalBidirectionalStream.Builder builder;
+  @Mock private BidirectionalStream.Builder builder;
   private final Object lock = new Object();
   private final TransportTracer transportTracer = TransportTracer.getDefaultFactory().create();
   private final Executor executor = new Executor() {
@@ -681,8 +680,8 @@ public final class CronetClientStreamTest {
             true,
             false);
     callback.setStream(stream);
-    ExperimentalBidirectionalStream.Builder getBuilder =
-        mock(ExperimentalBidirectionalStream.Builder.class);
+    BidirectionalStream.Builder getBuilder =
+        mock(BidirectionalStream.Builder.class);
     when(getFactory.newBidirectionalStreamBuilder(
             any(String.class), any(BidirectionalStream.Callback.class), any(Executor.class)))
         .thenReturn(getBuilder);
@@ -738,8 +737,8 @@ public final class CronetClientStreamTest {
             true,
             true);
     callback.setStream(stream);
-    ExperimentalBidirectionalStream.Builder builder =
-        mock(ExperimentalBidirectionalStream.Builder.class);
+    BidirectionalStream.Builder builder =
+        mock(BidirectionalStream.Builder.class);
     when(factory.newBidirectionalStreamBuilder(
             any(String.class), any(BidirectionalStream.Callback.class), any(Executor.class)))
         .thenReturn(builder);
@@ -770,8 +769,8 @@ public final class CronetClientStreamTest {
             true,
             true);
     callback.setStream(stream);
-    ExperimentalBidirectionalStream.Builder builder =
-        mock(ExperimentalBidirectionalStream.Builder.class);
+    BidirectionalStream.Builder builder =
+        mock(BidirectionalStream.Builder.class);
     when(factory.newBidirectionalStreamBuilder(
             any(String.class), any(BidirectionalStream.Callback.class), any(Executor.class)))
         .thenReturn(builder);
@@ -810,8 +809,8 @@ public final class CronetClientStreamTest {
             false,
             false);
     callback.setStream(stream);
-    ExperimentalBidirectionalStream.Builder builder =
-        mock(ExperimentalBidirectionalStream.Builder.class);
+    BidirectionalStream.Builder builder =
+        mock(BidirectionalStream.Builder.class);
     when(factory.newBidirectionalStreamBuilder(
             any(String.class), any(BidirectionalStream.Callback.class), any(Executor.class)))
         .thenReturn(builder);

--- a/cronet/src/test/java/io/grpc/cronet/CronetClientTransportTest.java
+++ b/cronet/src/test/java/io/grpc/cronet/CronetClientTransportTest.java
@@ -71,12 +71,7 @@ public final class CronetClientTransportTest {
   private MethodDescriptor<Void, Void> descriptor = TestMethodDescriptors.voidMethod();
   @Mock private ManagedClientTransport.Listener clientTransportListener;
   @Mock private BidirectionalStream.Builder builder;
-  private final Executor executor = new Executor() {
-      @Override
-      public void execute(Runnable r) {
-        r.run();
-      }
-    };
+  private final Executor executor = Runnable::run;
 
   @Before
   public void setUp() {

--- a/cronet/src/test/java/io/grpc/cronet/CronetClientTransportTest.java
+++ b/cronet/src/test/java/io/grpc/cronet/CronetClientTransportTest.java
@@ -17,7 +17,7 @@
 package io.grpc.cronet;
 
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertNotNull;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
@@ -91,7 +91,7 @@ public final class CronetClientTransportTest {
             false,
             false);
     Runnable callback = transport.start(clientTransportListener);
-    assertTrue(callback != null);
+    assertNotNull(callback);
     callback.run();
     verify(clientTransportListener).transportReady();
   }

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -23,10 +23,8 @@ auto-value-annotations = "com.google.auto.value:auto-value-annotations:1.10.4"
 checkstyle = "com.puppycrawl.tools:checkstyle:10.12.5"
 commons-math3 = "org.apache.commons:commons-math3:3.6.1"
 conscrypt = "org.conscrypt:conscrypt-openjdk-uber:2.5.2"
-# Update notes / 2023-07-19 sergiitk:
-#     Cronet (API and Embedded) upgrade is blocked by https://github.com/grpc/grpc-java/issues/10396.
-cronet-api = "org.chromium.net:cronet-api:108.5359.79"
-cronet-embedded = "org.chromium.net:cronet-embedded:108.5359.79"
+cronet-api = "org.chromium.net:cronet-api:119.6045.31"
+cronet-embedded = "org.chromium.net:cronet-embedded:119.6045.31"
 errorprone-annotations = "com.google.errorprone:error_prone_annotations:2.23.0"
 errorprone-core = "com.google.errorprone:error_prone_core:2.23.0"
 google-api-protos = "com.google.api.grpc:proto-google-common-protos:2.29.0"

